File: src/apps/daemon/internal/platform/security/encrypted_payload_envelope_test.go
package security

import (
	"encoding/json"
	"testing"
)

func TestEncryptDecryptPayload(t *testing.T) {
	passphrase := "whitevpn-daemon-secret-key"
	message := "https://relay.whitevpn.internal/catalog/nodes.json"

	encryptedJSON, err := EncryptPayload([]byte(message), passphrase)
	if err != nil {
		t.Fatalf("EncryptPayload failed: %v", err)
	}

	var env EncryptedPayloadEnvelope
	if err := json.Unmarshal([]byte(encryptedJSON), &env); err != nil {
		t.Fatalf("invalid json envelope: %v", err)
	}
	if env.Version != 1 || env.Algorithm != "AES-GCM" || env.Encoding != "base64url" {
		t.Fatalf("unexpected envelope headers: %+v", env)
	}

	decrypted, err := DecryptText(encryptedJSON, passphrase)
	if err != nil {
		t.Fatalf("DecryptText failed: %v", err)
	}
	if decrypted != message {
		t.Fatalf("expected '%s', got '%s'", message, decrypted)
	}
}

func TestDecryptWithWrongKeyFails(t *testing.T) {
	encryptedJSON, err := EncryptPayload([]byte("hello secret"), "key-one")
	if err != nil {
		t.Fatalf("EncryptPayload failed: %v", err)
	}

	_, err = DecryptText(encryptedJSON, "key-two")
	if err == nil {
		t.Fatalf("expected decryption error with wrong key")
	}
}

func TestParseAndEncryptIPList(t *testing.T) {
	rawText := "192.0.2.1  198.51.100.1\n\n192.0.2.1  bad.ip 256.0.0.1 1.1.1.1"
	ips := ParsePlaintextIPs(rawText)
	if len(ips) != 3 {
		t.Fatalf("expected 3 distinct valid IPs, got %d: %v", len(ips), ips)
	}

	passphrase := "ip-list-passphrase"
	encrypted, err := EncryptIPList(ips, passphrase)
	if err != nil {
		t.Fatalf("EncryptIPList failed: %v", err)
	}

	decryptedIPs, err := DecryptIPList(encrypted, passphrase)
	if err != nil {
		t.Fatalf("DecryptIPList failed: %v", err)
	}
	if len(decryptedIPs) != 3 || decryptedIPs[0] != "192.0.2.1" {
		t.Fatalf("unexpected decrypted IPs: %v", decryptedIPs)
	}
}

File: src/packages/lumicore/tests/encrypted_payload_envelope_test.rs
use lumicore::security::encrypted_payload_envelope::{
    EncryptedPayloadCodec, EncryptedPayloadEnvelope, EnvelopeError,
};

#[test]
fn test_encrypt_and_decrypt_plaintext() {
    let passphrase = "test-secret-passphrase-whitevpn";
    let message = "https://example.com/sub/token123?param=val";

    let encrypted_json = EncryptedPayloadCodec::encrypt(message.as_bytes(), passphrase)
        .expect("encryption failed");

    let envelope: EncryptedPayloadEnvelope = serde_json::from_str(&encrypted_json)
        .expect("json parse failed");
    assert_eq!(envelope.version, 1);
    assert_eq!(envelope.algorithm, "AES-GCM");
    assert_eq!(envelope.encoding, "base64url");
    assert!(!envelope.iv.is_empty());
    assert!(!envelope.ciphertext.is_empty());

    let decrypted = EncryptedPayloadCodec::decrypt_text(&encrypted_json, passphrase)
        .expect("decryption failed");
    assert_eq!(decrypted, message);
}

#[test]
fn test_wrong_passphrase_fails_decryption() {
    let encrypted_json = EncryptedPayloadCodec::encrypt(b"secret payload", "correct-key")
        .unwrap();

    let err = EncryptedPayloadCodec::decrypt(&encrypted_json, "wrong-key")
        .expect_err("decryption should fail with wrong key");

    match err {
        EnvelopeError::CryptoError(_) => {}
        other => panic!("expected CryptoError, got {:?}", other),
    }
}

#[test]
fn test_parse_and_encrypt_ip_list() {
    let raw_ips = "192.0.2.1   198.51.100.1\n203.0.113.5\n\n192.0.2.1\ninvalid-ip  999.999.999.999\n1.1.1.1";
    let parsed = EncryptedPayloadCodec::parse_plaintext_ips(raw_ips);
    // Should deduplicate and keep valid IPv4
    assert_eq!(parsed, vec!["192.0.2.1", "198.51.100.1", "203.0.113.5", "1.1.1.1"]);

    let passphrase = "secure-clean-ip-key";
    let encrypted = EncryptedPayloadCodec::encrypt_ip_list(
        &["1.1.1.1", "8.8.8.8", "1.0.0.1"],
        passphrase,
    ).unwrap();

    let decrypted_ips = EncryptedPayloadCodec::decrypt_ip_list(&encrypted, passphrase)
        .unwrap();
    assert_eq!(decrypted_ips, vec!["1.1.1.1", "8.8.8.8", "1.0.0.1"]);
}
